### PR TITLE
Assert `DataType::Index` in the ctor of `TensorIndex`

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: csrc/ops/utils.cpp [hunk 1]
```diff
@@ -314,7 +314,7 @@ Val* newValLike(Val* val, DataType dtype) {
     return newOutputTV({val}, dtype);
   }
 
-  return newScalar(vtype, dtype);
+  return newScalar(ValType::Others, dtype);
```
Comment (naoyam) [inline csrc/ops/utils.cpp:317]: Why this change?

Comment (zasdfgbnm) [reply csrc/ops/utils.cpp:317]: Otherwise we will be unable to do things like `maybeCastOp(TensorIndex)`

Comment (naoyam) [reply csrc/ops/utils.cpp:317]: Sorry, I'm lost here. Why is it? Is it documented somewhere?

Comment (zasdfgbnm) [reply csrc/ops/utils.cpp:317]: I don't think it is documented anywhere. It is just because the results of ops on `TensorIndex` are no longer a `TensorIndex` but a general `Val`. So we should not always preserve the same `ValType` here.

Comment (naoyam) [reply csrc/ops/utils.cpp:317]: OK, now it makes sense. I'd say ops on `TensorIndex` have been `ValType::Others`. It's just we were not using `maybeCastOp` before but manually created a new Val of type `ValType::Others`. 

Part of the reason I feel this is a little odd would be probably due to `TensorIndex`. I remember you mentioned that it should be an expression rather than a val, which I agree.

**Review by naoyam — APPROVED** [review]
LGTM
